distributed/worker.py

```diff
@@ -3089,7 +3093,8 @@ async def gather_dep(
                         self.batched_stream.send(
                             {"op": "missing-data", "errant_worker": worker, "key": d}
                         )
-                        recommendations[ts] = "fetch" if ts.who_has else "missing"
+                        if ts.state != "released":
+                            recommendations[ts] = "fetch" if ts.who_has else "missing"
```

Comment:
@fjetter @gjoseph92 what do you think about these changes?

We were seeing released -> missing transitions in the rechunk/killworker test (also in this branch now).  My guess is that we just want to keep things as released.  Is that correct in your view?

Reply:
The key must not be in `released` at this point. Do you have a story leading up to this?

Reply:
Couldn't reproduce, yet.

---

Keys are not allowed to be released at this point. This was one of the major sources of deadlocks last year that keys were released while either being fetched or executed that created horrible race conditions. That's why we have this weird cancelled state to ensure this cannot happen.

We should not add this condition in here, the finally clause of gather dep is likely the most buggy and fragile piece of code of our code base right now and I really don't want to add any other logic without a _reliable_ way of triggering this condition.

Reply:
> Couldn't reproduce, yet.

You may have to run the test a few times.  Also I agree with you about the oddities of erring in these methods.  They should percolate upwards to the entire worker, and out to the client.

Reply:
Running at scale with Coiled this fix does make the problem go away.